### ISSUE TITLE
Laplacian coordinate for each vertex

### ISSUE THREAD

**Issue by Lakmini** [pyg-team/pytorch_geometric#1792]
## ❓ Questions & Help

I'm new to pytorchGeometric and I'm currently working with volumetric meshes. I need to implement discrete Laplacian loss. For that I should first calculate the Laplacian coordinate for each vertex as shown below.

We can define a Laplacian coordinate for each vertex p as:

![laplace_loss_vertex_p](https://user-images.githubusercontent.com/8365715/97892157-92dfc280-1d27-11eb-8a46-d2de5ef3fa9e.png)

where p represents the 3D coordinates of a given vertex and the k represents the adjacent vertices to this given vertex.

I have gone through the **get_laplacian function** implemented in the utils package. This functionality outputs the Laplacian Matrix as the 1-D tensor (i.e. edge_weight) and the edge_index tensor which consists of the original edge index + self loops.

How can I incorporate the vertex coordinates to calculate the Laplacian coordinate for each vertex by utilizing this functionality ?
I'm asking this because when we are calculating this by hand we normally use a 2D matrix representation where diagonal elements represent the degree of the vertices, adjacent nodes represent by assigning -1 and otherwise we put zero (i.e. non-adjacent nodes). But the get_laplacian outputs the edge_weight as a 1-D tensor by concatenating the diagonal elements to the original -edge_weight.

I highly appreciate your willingness to help me out regarding this issue.

**Comment by rusty1s** [comment]
You can compute the Laplacian coordinate with `scatter_mean` (I don't think it's necessary to call `get_laplacian` here):
```python
from torch_scatter import scatter_mean

p = ...  # [N, ...]
edge_index = ... [2, E]
row, col = edge_index

delta = p - scatter_mean(col, p[row], dim=0, dim_size=p.size(0))
```

**Comment by Lakmini** [comment]
Thank you very much. But there should be a small change due to the parameter mismatch. The source points should be the first parameter and the index is the second one.

scatter_mean(p[row], col, dim=0, dim_size=p.size(0))

Thanks again.

**Comment by Lakmini** [comment]
I have one more issue. I need to do small modification as in the below equation. How can I obtain this from the scatter operation ?

![laplace_new_loss](https://user-images.githubusercontent.com/8365715/98006746-896b5e80-1dea-11eb-9ee7-952e5668416f.png)

Here vi indicates the coordinates of the given vertex and the vj represents the nodes adjacent to that given node.

**Comment by Lakmini** [comment]
I have implemented this as you can see below.

scatter_mean((p[row] - p[col]), col, dim=0, dim_size=p.size(0))

I think this is OK. right ?

**Comment by rusty1s** [comment]
That‘s correct :)

**Comment by Lakmini** [comment]
I need one more help from you. This is the last one. :)

 Once calculated the delta as mentioned at the beginning, now I want to do the below calculation.

![delta_pariwise_zigma](https://user-images.githubusercontent.com/8365715/98159803-1985d280-1ed5-11eb-9811-7fc2210d0121.png)

here delta_p represents the actual and delta_p_dash represents the predicted Laplacian coordinate of a vertex. In order to calculate this I just do the following.

```
delta_dash = predictedTensor - scatter_mean(predictedTensor[row], col, dim=0, dim_size=predictedTensor.size(0))
delta = targetTensor - scatter_mean(targetTensor[row], col, dim=0, dim_size=targetTensor.size(0))

pairwiseDist = nn.PairwiseDistance(p=2)
torch.sum(self.pairwiseDist(delta_dash, delta))
```

Additionally, I need to calculate the edge length regularization as in the below equation.

![edge_length_reg](https://user-images.githubusercontent.com/8365715/98160651-5ef6cf80-1ed6-11eb-805b-bfa77399ebb3.png)

where p represents the 3D coordinates of a given vertex and the k represents the adjacent vertices to this given vertex.

I have done this calculation as below.

```
row, col = edge_index
pairwiseDist = nn.PairwiseDistance(p=2)
torch.sum(pdist(predictedTensor[row], predictedTensor[col]))
```

**What I need to know is that whether these calculations are correct or not ?**
Please help me on this matter.

**Comment by rusty1s** [comment]
That looks absolutely correct to me.

**Comment by Lakmini** [comment]
Shouldn't I need to take pairwiseDist**2 since there is a superscript 2  in the equations. By default it is calculated square-root values ?

**Comment by rusty1s** [comment]
Yeah, that's right. I guess it doesn't really matter though :)

**Comment by Lakmini** [comment]
There is an issue with the way that I have calculated the first two equations. They should be corrected as below since we need to index by row and not with the column indexes. :)
We need to index by rows to preserve the order when we are subtracting from the original vertex coordinates. right ?

```
p - scatter_mean(p[col], row, dim=0, dim_size=p.size(0))

scatter_mean((p[row] - p[col]), row, dim=0, dim_size=p.size(0))

```

**Comment by rusty1s** [comment]
Aren't both formulations equivalent?

**Comment by Lakmini** [comment]
Yes. Both formulations as indicated below calculate exactly the same results. But, if I index with column indices then the results belong to each formulation (I have used only the first formulations to test) are wrong. So the second parameter would be row index tensor.

```
p - scatter_mean(p[col], row, dim=0, dim_size=p.size(0))
scatter_mean((p[row] - p[col]), row, dim=0, dim_size=p.size(0))
```
